### PR TITLE
CR-1214 Webform request sent to uk.gov.service.notify

### PR DESCRIPTION
**Pull request by MartinHumphrey**
# Motivation and Context
Receive a Webform event message and send to the GOV.UK Notify service to send an email with the details from the event message.

# What has changed
New inbound event flow configured for Webform events. New WebformEventReceiver message endpoint added and WebformService extended to send a Webform email request to the GOV.UK Notify service.

# How to test?
Unit tests added for new functionality. Run successfully against latest census RH cucumber tests as new functionality not affecting existing tests.
On deployment to census-rh-dev environment queues notify.rh.webform and notify.rh.webform.dlq were created in RabbitMQ successfully automatically.

# Links
JIRA CR-1214 https://collaborate2.ons.gov.uk/jira/browse/CR-1214
Previous PR to place message on Rabbit queue: https://github.com/ONSdigital/census-rh-service/pull/91
GOV.UK Notify Java client documentation https://docs.notifications.service.gov.uk/java.html#java-client-documentation

### CODE REVIEW

File: src/main/java/uk/gov/ons/ctp/integration/rhsvc/config/InboundEventIntegrationConfig.java [hunk 1]
```diff
@@ -211,16 +256,6 @@ private Jackson2JsonMessageConverter jsonMessageConverter(
     return converter;
   }
 
-  @Bean
```
Comment (rpcatling) [inline src/main/java/uk/gov/ons/ctp/integration/rhsvc/config/InboundEventIntegrationConfig.java:214]: thanks for removing these. these were in the XML, and I forgot to remove them later when I noticed they weren't used.

File: src/main/java/uk/gov/ons/ctp/integration/rhsvc/service/impl/WebformServiceImpl.java [hunk 2]
```diff
@@ -23,4 +44,46 @@ public String sendWebformEvent(Webform webform) {
...
+          .with("templateId", response.getTemplateId().toString())
+          .with("templateVersion", response.getTemplateVersion())
+          .debug("Gov Notify sendEmail response recieved");
+    } catch (NotificationClientException ex) {
+      log.with("reference", reference)
+          .with("webform", webform)
+          .with("emailToAddress", emailToAddress)
+          .with("status", ex.getHttpResult())
+          .with("message", ex.getMessage())
+          .error("Gov Notify sendEmail NotificationClientException error");
```
[the remark this answers is not in the document]
Comment (rpcatling) [reply src/main/java/uk/gov/ons/ctp/integration/rhsvc/service/impl/WebformServiceImpl.java:75]: Don't forget to add new ERROR log details in the JIRA and inform Akriti for her dashboard work.

**Review by rpcatling — CHANGES_REQUESTED** [review]
Looks pretty good. A few minor comments, as well as the agreed change to customise retry.

File: src/main/java/uk/gov/ons/ctp/integration/rhsvc/service/impl/WebformServiceImpl.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (MartinHumphrey) [reply src/main/java/uk/gov/ons/ctp/integration/rhsvc/service/impl/WebformServiceImpl.java:75]: Changed to add stack trace as per convention, class type removed from message.

Updated CR-1214 JIRA and also CR-1280 which covered sending of Webform message to queue. Discussed with Akriti.

File: src/main/java/uk/gov/ons/ctp/integration/rhsvc/config/InboundEventIntegrationConfig.java (hunk 1 above)
Comment (MartinHumphrey) [reply src/main/java/uk/gov/ons/ctp/integration/rhsvc/config/InboundEventIntegrationConfig.java:214]: No problem.

File: src/main/java/uk/gov/ons/ctp/integration/rhsvc/service/impl/WebformServiceImpl.java [hunk 3]
```diff
@@ -1,20 +1,41 @@
...
+  private static final String TEMPLATE_EMAIL = "respondent_email";
+  private static final String TEMPLATE_REGION = "respondent_region";
+  private static final String TEMPLATE_CATEGORY = "respondent_category";
+  private static final String TEMPLATE_DESCRIPTION = "respondent_description";
+
   @Autowired private EventPublisher eventPublisher;
 
+  @Autowired private NotificationClientApi notificationClient;
+
+  @Autowired private AppConfig appConfig;
```
[the remark this answers is not in the document]
Comment (MartinHumphrey) [reply src/main/java/uk/gov/ons/ctp/integration/rhsvc/service/impl/WebformServiceImpl.java]: Spring does now [advocate constructor-based injection](https://docs.spring.io/spring-framework/docs/4.3.x/spring-framework-reference/html/beans.html#beans-setter-injection) it seems. Have replaced with constructor injection.

File: src/main/java/uk/gov/ons/ctp/integration/rhsvc/event/impl/WebformEventReceiverImpl.java [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+import uk.gov.ons.ctp.integration.rhsvc.event.WebformEventReceiver;
+import uk.gov.ons.ctp.integration.rhsvc.service.WebformService;
+
+/** Service implementation responsible for receipt of Webform Events. */
+@MessageEndpoint
+public class WebformEventReceiverImpl implements WebformEventReceiver {
+
+  private static final Logger log = LoggerFactory.getLogger(WebformEventReceiverImpl.class);
+
+  @Autowired private WebformService webformService;
```
[the remark this answers is not in the document]
Comment (MartinHumphrey) [reply src/main/java/uk/gov/ons/ctp/integration/rhsvc/event/impl/WebformEventReceiverImpl.java]: Have changed to constructor injection.

File: src/main/java/uk/gov/ons/ctp/integration/rhsvc/config/InboundEventIntegrationConfig.java [hunk 5]
```diff
@@ -132,6 +133,26 @@ public SimpleMessageListenerContainer uacEventListenerContainer(
...
+   *
+   * @param connectionFactory connection factory
+   * @param eventRetryAdvice retry advice
+   * @param rabbitDownBackOff backoff for when rabbit problems occur
+   * @return listener container for Webform events.
+   */
+  @Bean
+  public SimpleMessageListenerContainer webformEventListenerContainer(
+      ConnectionFactory connectionFactory,
+      StatelessRetryOperationsInterceptorFactoryBean eventRetryAdvice,
```
[the remark this answers is not in the document]
Comment (MartinHumphrey) [reply src/main/java/uk/gov/ons/ctp/integration/rhsvc/config/InboundEventIntegrationConfig.java]: Have separated out Webform SimpleMessageListenerContainer from UAC, Case configuration.

**Review by rpcatling — APPROVED** [review]

**Review by philwhiles — APPROVED** [review]

**Pull request merged**